src/hotspot/share/prims/jvmtiEnv.cpp

```diff
@@ -1190,9 +1190,15 @@ JvmtiEnv::StopThread(jthread thread, jobject exception) {
 
   jvmtiError err = get_threadOop_and_JavaThread(tlh.list(), thread, &java_thread, &thread_oop);
 
-  if (thread_oop != nullptr && thread_oop->is_a(vmClasses::BaseVirtualThread_klass())) {
-    // No support for virtual threads (yet).
-    return JVMTI_ERROR_UNSUPPORTED_OPERATION;
+  bool is_virtual = thread_oop != nullptr && thread_oop->is_a(vmClasses::BaseVirtualThread_klass());
+
+  if (is_virtual && !is_JavaThread_current(java_thread, thread_oop)) {
+    if (!JvmtiVTSuspender::is_vthread_suspended(thread_oop)) {
+      return JVMTI_ERROR_THREAD_NOT_SUSPENDED;
```

Comment:
Does JvmtiVTSuspender::is_vthread_suspended work for the alternative virtual thread implementation (-XX:+VMContinuations)?

Reply:
Nice catch - fixed now.

Reply:
Okay but JvmtiVTMSTransitionDisabler prevent the thread from being resumed for the -XX:-VMContinuations case?

Reply:
Thank you for the catch. Will check it. I have to extend the test to cover the BoundVirtualThread case enabled with the flag `-XX:-VMContinuations`.

Reply:
The scenario that I'm wondering about is where a virtual thread is resumed at around the same time that JVMTI StopThread is called. Not easy to test.

Reply:
Seems we should have a stress test for that.

Reply:
> The scenario that I'm wondering about is where a virtual thread is resumed
> at around the same time that JVMTI StopThread is called.

This kind of scenario is not typical.
The debugger should keep virtual thread suspended when making a call to JVMTI StopThread.
Normally, this kind of race should never happened with the JDWP agent.
Chris, why do you think it is important to have a stress test for this?
Also, do you have any testing scenario in mind?

Reply:
I guess if it is not something that would typically ever be done in realistic situations, then there is no need for a stress test.

Reply:
Okay, thanks.
I've extended the test to cover BoundVirtualThread's as well.